packages/report-server/src/__tests__/reportBuilder/transform/insert.test.ts

```diff
@@ -0,0 +1,215 @@
+      { dataElementValue: 2 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { dataElementValue: 5 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+    ]);
+  });
+
+  it('can insert new rows at the beginning of the list', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+        position: 'start',
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { dataElementValue: 4 },
+      { dataElementValue: 2 },
+      { dataElementValue: 5 },
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+    ]);
+  });
+
+  it('can insert specific new rows using a where clause', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        "'dataElementValue'": '$row.value',
+        position: 'after',
+        where: "not(eq($row.period, '20200101'))",
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { dataElementValue: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+      { dataElementValue: 5 },
+    ]);
+  });
+
+  // USE CASES
+  it('can sum all values into a new totals row', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        where: 'eq($index, length($table) - 1)',
+        "'Total'": 'sum($all.value)',
+      },
+    ]);
+    expect(transform(MULTIPLE_ANALYTICS)).toEqual([
+      { period: '20200101', organisationUnit: 'TO', dataElement: 'BCD1', value: 4 },
+      { period: '20200102', organisationUnit: 'TO', dataElement: 'BCD1', value: 2 },
+      { period: '20200103', organisationUnit: 'TO', dataElement: 'BCD1', value: 5 },
+      { Total: 11 }
+    ]);
+  });
+
+  it('compare adjacent rows to insert between', () => {
+    const transform = buildTransform([
+      {
+        transform: 'insert',
+        position: 'before',
+        where: 'and(exists($previous.organisationUnit), not(eq($previous.organisationUnit, $row.organisationUnit)))',
+        "'Total_BCD1'": "sum($where(f($otherRow) = eq($otherRow.organisationUnit, $previous.organisationUnit)).BCD1)",
+        "'Total_BCD2'": "sum($where(f($otherRow) = eq($otherRow.organisationUnit, $previous.organisationUnit)).BCD2)",
+      },
+      {
+        transform: 'insert',
+        where: 'eq($index, length($table) - 1)',
+        "'Total_BCD1'": 'sum($all.BCD1) - sum($all.Total_BCD1)',
+        "'Total_BCD2'": 'sum($all.BCD2) - sum($all.Total_BCD2)',
+      },
+    ]);
```

Comment:
This is crying out for a `$next`...but you've already added a heap of awesome stuff, and it is possible just using `$previous` as you've demonstrated, so happy keeping things restricted for now

Reply:
I wrote the first insert here before I added the extra lookups. I suppose you could do a next lookup as `$table[$index + 1]` if you really wanted to